File: leetcode/lib/arrays_hashing/242_valid_anagram.rb
# Given two strings s and t, return true if t is an anagram of s, and false otherwise.
# An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

=begin
time: O(s + t)
space: O(s) (technically, will only ever be 26 characters is only using lowercase alphabet)
create a hash map from s, where key = letter and value = num occurrences	
iterate through t, decrementing each letter when found
return true if hash map is empty
return false if hash map has any values

time: O((s + t) * (s * log s))
space: O(1) (if sorting doesn't take extra space (sometimes they take O(s)))
sort s and t
return s_sorted == t_sorted
=end

def is_anagram(s, t)
	return false if s.length != t.length

	letter_nums = Hash.new(0)

	s.each_char do |char|
		letter_nums[char] += 1
	end

	t.each_char do |char|
		if letter_nums.key?(char)
			if letter_nums[char] == 1
				letter_nums.delete(char)
			else
				letter_nums[char] -= 1
			end
		else
			return false
		end
	end

	if letter_nums.empty?
		return true
	else
		return false
	end
end

puts("--- TEST 1 ---")
s = "anagram"
t = "nagaram"
puts(is_anagram(s, t))
# expect true

puts("--- TEST 2 ---")
s = "rat"
t = "car"
puts(is_anagram(s, t))
# expect false

File: leetcode/lib/arrays_hashing/242_valid_anagram.py
"""
Given two strings s and t, return true if t is an 
anagram of s, and false otherwise.
"""

"""
Add all letters of s into map (letter -> freq)
Iterate over t:
    If letter not in map, return false
    If letter in map, decrement
    If letter in map == 1, remove from map
If map is empty, return True. Else, False.

Time: O(s + t)
Space: O(s)
"""

from utils import test
from collections import defaultdict


class Solution:
    def is_anagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        letters = defaultdict(int)

        for letter in s:
            letters[letter] += 1

        for letter in t:
            if letter not in letters:
                return False
            else:
                if letters[letter] == 1:
                    del letters[letter]
                else:
                    letters[letter] -= 1

        if len(letters.keys()) == 0:
            return True
        else:
            return False

    def is_anagram2(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        count_s = defaultdict(int)
        count_t = defaultdict(int)

        for i in range(len(s)):
            count_s[s[i]] += 1
            count_t[t[i]] += 1

        return count_s == count_t

    def is_anagram_sort(self, s: str, t: str) -> bool:
        """
        Time: O(s log s + t log t)
        Space: O(1) or O(n + m) (depending on sorting alg)
        """
        if len(s) != len(t):
            return False

        return sorted(s) == sorted(t)


def main():
    solution = Solution()
    test(1, True, solution.is_anagram2("anagram", "nagaram"))
    test(2, False, solution.is_anagram2("rat", "car"))
    test(3, False, solution.is_anagram2("ab", "a"))


if __name__ == "__main__":
    main()
